syntax = "proto3";

message Todo {
    int32 id = 1;
    int32 user_id = 2;
    string title = 3;
    bool completed = 4;
}

message AddTodoRequest {
    string auth_token = 1;
    Todo todo = 2;
}

message AddTodoResponse {
    Todo todo = 1;
}

message RemoveTodoRequest {
    string auth_token = 1;
    Todo todo = 2;
}

message RemoveTodoResponse {
    enum Result {
        SUCCESS = 0;
        FAILURE = 1;
    }
    Result result = 2;
    string message = 3;
}

message FetchTodosRequest {
    string auth_token = 1;
    int32 page_number = 2;
    int32 page_size = 3;
}

message FetchTodosResponse {
    repeated Todo todos = 1;
}

message UpdateTodoRequest {
    string auth_token = 1;
    Todo todo = 2;
}

message UpdateTodoResponse {
    Todo todo = 1;
}

service TodoList {
    rpc AddTodo (AddTodoRequest) returns (AddTodoResponse) {}
    rpc RemoveTodo (RemoveTodoRequest) returns (RemoveTodoResponse) {}
    rpc FetchTodos (FetchTodosRequest) returns (FetchTodosResponse) {}
    rpc UpdateTodo (UpdateTodoRequest) returns (Todo) {}
}
